Improve z-wave thermostat support

oandrew opened the pull request:
Discover thermostats using `COMMAND_CLASS_THERMOSTAT_MODE` so that it is a single entity when there are multiple setpoints. 
z-wave docs mention this command class is required to be present.
Add support for single/dual target temperature depending
on the current thermostat mode.
I also verified thermostat is correctly exposed through Amazon Alexa/Google Assistant and single/dual setpoints are dynamically updating based on the current mode.

## Breaking Change:
This changes the  primary command class for z-wave thermostats in discovery schemas from `COMMAND_CLASS_THERMOSTAT_SETPOINT` to `COMMAND_CLASS_THERMOSTAT_MODE`.
This will cause a typical dual setpoint thermostat to be correctly represented as a single entity.

## Description:
Previously z-wave thermostats were discovered using `COMMAND_CLASS_THERMOSTAT_SETPOINT` as a primary value.
That caused a single thermostat to be exposed multiple times - once for each setpoint. 
Also this PR adds dynamic single/dual setpoint support based on the current thermostat mode.
e.g. if current mode is `HVAC_MODE_HEAT` or `HVAC_MODE_COOL`  - expose single target temp `ATTR_TEMPERATURE`, if current mode is `HVAC_MODE_HEAT_COOL` - expose target range attributes `ATTR_TARGET_TEMP_LOW` and `ATTR_TARGET_TEMP_HIGH`

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>
I will create PR for docs once I get feedback on this PR

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

oandrew committed:
Improve z-wave thermostat support

Discover thermostat using COMMAND_CLASS_THERMOSTAT_MODE
so that it is a single entitiy in case there are multiple
setpoints. z-wave docs mention it is always present.
Add support for single/range target temperature depending
on the current thermostat mode.

cgarwood commented on homeassistant/components/zwave/climate.py:
```diff
@@ -430,12 +458,36 @@ def target_temperature(self):
+            (setpoint_low, setpoint_high) = current_setpoints
+            target_temp_low = kwargs.get(ATTR_TARGET_TEMP_LOW)
+            target_temp_high = kwargs.get(ATTR_TARGET_TEMP_HIGH)
+            print("SETTING: ", target_temp_low, target_temp_high)
```
`print` statement should be removed or replaced with a `_LOGGER` call.

cgarwood commented:
Z-Wave side looks fine to me, but I'd like to have someone with more knowledge of the `climate` platforms take a look

oandrew committed:
Remove debug print

frenck commented:
Thank you for your contribution thus far! 🎖 Since this is a significant contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️ 

Please, add your GitHub username to the `manifest.json` of this integration.

For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

MartinHjelmare commented on homeassistant/components/zwave/climate.py:
```diff
@@ -154,10 +161,27 @@ def __init__(self, values, temp_unit):
-        support = SUPPORT_TARGET_TEMPERATURE
+        support = 0
+        current_setpoints = self._current_mode_setpoints()
+        if len(current_setpoints) == 1:
```
Can `current_setpoints` change during entity lifetime? Supported features are not allowed to change during entity lifetime.

oandrew replied:
Yep. It depends on the current mode currently.
Let me check how UI/alexa/google components interact with climate component.
My main concern is exposing the correct number of setpoints based on the current mode.

MartinHjelmare replied:
The target temperature properties should return `None` if the current mode isn't using a corresponding target temperature. See smartthings climate platform eg:
https://github.com/home-assistant/home-assistant/blob/7d2a8b81370fdf643ce059e8f53d831de442eeaa/homeassistant/components/smartthings/climate.py#L297-L323

oandrew replied:
Makes sense. I'll update a PR. Thanks!

adrianschroeter replied:
There are also thermostats (like Eurotronic Spirit-Z) which have two heating setpoints. One for normal and one for Eco mode. So this code should maybe not be limited to cooling?

oandrew replied:
I've updated the PR to address the comments

Santobert commented on homeassistant/components/zwave/climate.py:
```diff
@@ -447,19 +498,19 @@ def set_fan_mode(self, fan_mode):
         operation_mode = self._hvac_mapping.get(hvac_mode)
         _LOGGER.debug("Set operation_mode to %s", operation_mode)
-        self.values.mode.data = operation_mode
+        self.values.primary.data = operation_mode
```
Please ignore that if it's a bad idea. We can prevent the setting of a mode that is already active with something like the following function:

```python
    def update_mode(self, operation_mode):
        """Update the operation mode only if it is not already set."""
        if self.values.primary.data != operation_mode:
            self.values.primary.data == operation_mode
```

I noticed that my zwave thermostats move when an active mode is activated again.

oandrew committed:
Refactor Z-Wave dynamic setpoint(s) selection

- use explicit mapping between modes and setpoints as defined in
Z-Wave specs
- add tests for away (2 setpoints) and heat eco (1 setpoint) modes

Santobert commented on homeassistant/components/zwave/climate.py:
```diff
@@ -66,6 +73,24 @@
+    "dry air": ("setpoint_dry_air",),
+    "moist air": ("setpoint_moist_air",),
+    "auto changeover": ("setpoint_auto_changeover",),
+    "heat econ": ("setpoint_eco_heating",),
```
The eco mode is called "Heat Eco" in case of Eurotronic Spirit-Z thermostats. I'm not sure if this makes a difference.

oandrew replied:
Thanks for noticing that.
The  labels I used are from https://github.com/OpenZWave/open-zwave/blob/85ca5769a57d13f328e65b415d95e95647994366/cpp/src/command_classes/ThermostatMode.cpp#L56-L94 .
But it turns out open-zwave let's you override the labels in xml as well.
In your case here: https://github.com/OpenZWave/open-zwave/blob/85ca5769a57d13f328e65b415d95e95647994366/config/eurotronic/eur_spiritz.xml#L71
So indexes are the same but labels are different :(
I just ran grep across all XML configs and found various confusing labels for same indexes e.g.
`Energy Heat = Energy Saving = Heat Eco = Heat Econ` and so on.

I will add the aliases I've found.
Sometime later it makes sense to switch to indexes instead of aliases since they seem to be more stable

oandrew committed:
Add non-standard thermostat mode aliases

oandrew commented:
Could someone review this when you have a chance? Thanks

Santobert commented:
It'd be great if we could continue here. I'm really looking forward to seeing these changes in the next release.
@oandrew Could you please improve the code coverage by writing some more tests. Besides some single lines only the aux_heat methods are untested.

balloob reviewed (approved):
This looks great! Thanks
